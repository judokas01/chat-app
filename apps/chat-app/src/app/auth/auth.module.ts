import { Module } from '@nestjs/common'
import { LoginService } from './login/login.service'
import { LogoutService } from './logout/logout.service'
import { RegisterService } from './register/register.service'
import { AuthenticateService } from './authenticate/authenticate.service'

@Module({
    providers: [LoginService, LogoutService, RegisterService, AuthenticateService],
})
export class AuthModule {}
