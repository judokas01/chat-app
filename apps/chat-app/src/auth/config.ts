export const AUTH_MODULE_SALT_ROUNDS = 10
