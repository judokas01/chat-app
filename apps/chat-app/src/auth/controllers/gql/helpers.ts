import { LoginArgsGql } from './request-type'

export const getLogInGqlRequest = (
    args: LoginArgsGql,
): { query: string; variables: LoginArgsGql } => {
    return {
        query: `
            query ($password:String!,$userName:String!) {
                login(password: $password, userName: $userName) {
                    accessToken
                    accessToken
                }
            }`,
        variables: args,
    }
}

export const commonUtil = <
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    T extends new (...args: any[]) => any,
    M extends keyof InstanceType<T>,
    A extends Parameters<InstanceType<T>[M]>,
>(args: {
    type: 'query' | 'mutation'
    name: string
    class: T
    classMethod: M
    args: A
}) => {
    const methodArguments = args.args.at(0)
    const inferredArgs = inferArguments(methodArguments)

    const returnQuery: Awaited<ReturnType<InstanceType<T>[M]>> = {}

    const queryStart = args.type === 'query' ? 'query' : 'mutation'

    const queryString = `${queryStart}
    (${inferredArgs.map(({ key, type }) => `$${key}:${type}`).join(', ')})
    {
        ${args.name}(${inferredArgs.map(({ key }) => `${key}:$${key}`).join(', ')}) {
            accessToken
            renewToken
        }
    } 
    `

    return { variables: methodArguments, [args.type]: queryString }
}

const inferArguments = (methodArguments: unknown) => {
    const resolvedArguments = []
    if (isObject(methodArguments)) {
        for (const key in methodArguments) {
            resolvedArguments.push({ key, type: typeToGqlType(typeof methodArguments[key]) })
        }
    }

    return resolvedArguments
}

const isObject = (obj: unknown): obj is Record<string, unknown> => {
    return typeof obj === 'object' && obj !== null
}

const typeToGqlType = (type: string) => {
    switch (type) {
        case 'string':
            return 'String!'
        case 'number':
            return 'Int!'
        case 'boolean':
            return 'Boolean!'
        default:
            return 'String!'
    }
}
