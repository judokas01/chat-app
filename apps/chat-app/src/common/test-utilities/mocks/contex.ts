/* eslint-disable @typescript-eslint/no-explicit-any */
import { ContextType, ExecutionContext } from '@nestjs/common'
import { RpcArgumentsHost, WsArgumentsHost } from '@nestjs/common/interfaces'

export const getMockContext = (overrides: { authToken?: string }): ExecutionContext => ({
    getArgByIndex: <T = any>(_index: number): T => {
        throw new Error('Function not implemented.')
    },
    getArgs: <T extends any[] = any[]>(): T => {
        return null as any
    },
    getClass: <T = any>(): T => {
        throw new Error('Function not implemented.')
    },
    getHandler: () => {
        throw new Error('Function not implemented.')
    },
    getType: <TContext extends string = ContextType>(): TContext => {
        return null as any
    },
    switchToHttp: () => ({
        getNext: <T = any>(): T => {
            throw new Error('Function not implemented.')
        },
        getRequest: <Request>(): Request => {
            const req: Request = {
                headers: {
                    get: () => `Bearer ${overrides.authToken ?? ''}`,
                },
            } as Request

            return req
        },
        getResponse: <T = any>(): T => {
            throw new Error('Function not implemented.')
        },
    }),
    switchToRpc: function (): RpcArgumentsHost {
        throw new Error('Function not implemented.')
    },
    switchToWs: function (): WsArgumentsHost {
        throw new Error('Function not implemented.')
    },
})
