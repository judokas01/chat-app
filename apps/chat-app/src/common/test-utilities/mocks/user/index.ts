import { UserInput } from '@root/common/entities/user.entity'
import { faker } from '@faker-js/faker'

const createRandomUserInput = (overrides?: Partial<UserInput>): UserInput => ({
    email: faker.internet.email(),
    password: faker.internet.password(),
    userName: faker.internet.userName(),
    ...overrides,
})

export const userMock = {
    random: {
        getOne: createRandomUserInput,
    },
}
