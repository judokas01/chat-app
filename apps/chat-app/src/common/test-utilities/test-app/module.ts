import { Test, TestingModule } from '@nestjs/testing'
import { PrismaService } from '@root/infrastructure/prisma/prisma.service'
import { UserPrismaRepository } from '@root/auth/repository/prisma/user.repository'
import { IUserRepository } from '@root/auth/repository/user-repository.interface'
import { JWT } from '@root/auth/common/jwt.module'
import { DynamicModule, Provider, ValidationPipe, Type } from '@nestjs/common'
import { ConfigService } from '../../config/config-service.service'
import { cleanDb, getRepositories } from './common'

export const getTestModule = async (args: {
    imports?: DynamicModule[]
    providers?: Provider[]
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    controllers?: Type<any>[]
}) => {
    const { imports, providers, controllers } = args
    const module: TestingModule = await Test.createTestingModule({
        controllers: [...(controllers ?? [])],
        imports: [JWT, ...(imports ?? [])],
        providers: [
            { provide: IUserRepository, useClass: UserPrismaRepository },
            PrismaService,
            ConfigService,
            ValidationPipe,
            ...(providers ?? []),
        ],
    }).compile()

    return { cleanDb: cleanDb(module), module, repositories: getRepositories(module) }
}

export type TestModule = Awaited<ReturnType<typeof getTestModule>>
