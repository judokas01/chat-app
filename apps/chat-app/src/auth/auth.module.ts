import { Module } from '@nestjs/common'
import { LoginService } from './login/login.service'
import { LogoutService } from './logout/logout.service'
import { RegisterService } from './register/register.service'
import { AuthenticateService } from './authenticate/authenticate.service'
import { JWT } from './common/jwt.module'

@Module({
    imports: [JWT],
    providers: [LoginService, LogoutService, RegisterService, AuthenticateService],
})
export class AuthModule {}
