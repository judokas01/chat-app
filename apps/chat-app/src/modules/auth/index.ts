import { Module, ValidationPipe } from '@nestjs/common'
import { ConfigService } from '@root/common/config/config-service.service'
import { PrismaService } from '@root/infrastructure/prisma/prisma.service'
import { LoginService } from './login/login.service'
import { RegisterService } from './register/register.service'
import { JwtAuthenticateService } from './authenticate/services/jwt-authenticate.service'
import { JWT } from './common/jwt.module'
import { RestController } from './controllers/rest/rest.controller'
import { IUserRepository } from './repository/user-repository.interface'
import { UserPrismaRepository } from './repository/prisma/user.repository'
import { IAuthenticateService } from './authenticate/authenticate.interface'
import { AuthResolver } from './controllers/gql/auth.resolver'

@Module({
    controllers: [RestController],
    exports: [
        LoginService,
        RegisterService,
        JwtAuthenticateService,
        ConfigService,
        PrismaService,
        ValidationPipe,
        AuthResolver,
        { provide: IUserRepository, useClass: UserPrismaRepository },
        { provide: IAuthenticateService, useClass: JwtAuthenticateService },
    ],
    imports: [JWT],
    providers: [
        LoginService,
        RegisterService,
        JwtAuthenticateService,
        ConfigService,
        PrismaService,
        ValidationPipe,
        AuthResolver,
        { provide: IUserRepository, useClass: UserPrismaRepository },
        { provide: IAuthenticateService, useClass: JwtAuthenticateService },
    ],
})
export class AuthModule {}
