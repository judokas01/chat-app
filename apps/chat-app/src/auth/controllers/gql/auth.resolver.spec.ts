import {
    TestInterfaceModule,
    getTestModuleWithInterface,
} from '@root/common/test-utilities/test-app/interface'
import { AuthModule } from '@root/auth/auth.module'
import { userMock } from '@root/common/test-utilities/mocks/user'
import { AuthResolver } from './auth.resolver'
import { getLogInGqlRequest, commonUtil } from './helpers'

describe('AuthResolver', () => {
    let resolver: AuthResolver
    let testModule: TestInterfaceModule

    beforeAll(async () => {
        testModule = await getTestModuleWithInterface({
            module: AuthModule,
            providers: [AuthResolver],
        })

        resolver = testModule.module.get<AuthResolver>(AuthResolver)

        type a = Parameters<AuthResolver['logIn']>
        type b = Awaited<ReturnType<AuthResolver['logIn']>>
        const a: a = [{ password: '123', userName: 'test' }]
        const b: b = { accessToken: '', renewToken: '' }
    })

    afterAll(async () => {
        await testModule.destroy()
    })

    beforeEach(async () => {
        await testModule.cleanDb()
    })

    it('should be defined', () => {
        expect(resolver).toBeDefined()
    })

    it('should get invalid credentials error message', async () => {
        const res = await testModule.requestGql.post('').send({
            query: `
            query ($password:String!,$userName:String!) {
                login(password: $password, userName: $userName) {
                    accessToken
                    accessToken
                }
            }`,
            variables: { password: 'asd', userName: 'asfasf' },
        })

        expect(res.body).toMatchSnapshot()
    })

    it('should get invalid credentials error message with test util', async () => {
        const user = userMock.random.getOne()

        const original = getLogInGqlRequest(user)
        const kk = commonUtil({
            args: [{ password: '123', userName: 'test' }],
            class: AuthResolver,
            classMethod: 'logIn',
            name: 'login',
            returnQuery: { accessToken: '', renewToken: '' },
            type: 'query',
        })

        console.log({ kk, original })

        const res = await testModule.requestGql.post('').send(kk)

        expect(res.body).toMatchSnapshot()
    })
})
