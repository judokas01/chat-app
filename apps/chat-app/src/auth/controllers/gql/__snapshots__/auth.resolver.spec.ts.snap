// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthResolver should get invalid credentials error message 1`] = `
{
  "data": null,
  "errors": [
    {
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
      },
      "locations": [
        {
          "column": 17,
          "line": 3,
        },
      ],
      "message": "Invalid username and password combination.",
      "path": [
        "login",
      ],
    },
  ],
}
`;

exports[`AuthResolver should get invalid credentials error message with test util 1`] = `
{
  "data": null,
  "errors": [
    {
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
      },
      "locations": [
        {
          "column": 9,
          "line": 4,
        },
      ],
      "message": "Invalid username and password combination.",
      "path": [
        "login",
      ],
    },
  ],
}
`;
